src/server/saved_objects/migrations/core/index_migrator.ts

```diff
@@ -0,0 +1,179 @@
+
+/*
+ * Core logic for migrating the mappings and documents in an index.
+ */
+export class IndexMigrator {
+  private opts: MigrationOpts;
+
+  /**
+   * Creates an instance of IndexMigrator.
+   *
+   * @param {MigrationOpts} opts
+   */
+  constructor(opts: MigrationOpts) {
+    this.opts = opts;
+  }
+
+  /**
+   * Migrates the index, or, if another Kibana instance appears to be running the migration,
+   * waits for the migration to complete.
+   *
+   * @returns {Promise<MigrationResult>}
+   */
+  public async migrate(): Promise<MigrationResult> {
+    const context = await migrationContext(this.opts);
+
+    return coordinateMigration({
+      log: context.log,
+
+      pollInterval: context.pollInterval,
+
+      async isMigrated() {
+        const action = await requiredAction(context);
+        return action === MigrationAction.None;
+      },
+
+      async runMigration() {
+        const action = await requiredAction(context);
+
+        if (action === MigrationAction.None) {
+          return { status: 'skipped' };
+        }
+
+        if (action === MigrationAction.Patch) {
+          return patchSourceMappings(context);
+        }
+
+        return migrateIndex(context);
+      },
+    });
+  }
+}
+
+/**
+ * Determines what action the migration system needs to take (none, patch, migrate).
+ */
+async function requiredAction(context: Context): Promise<MigrationAction> {
+  const { callCluster, alias, documentMigrator, dest } = context;
+
+  const hasMigrations = await Index.hasMigrations(
+    callCluster,
+    alias,
+    documentMigrator.migrationVersion
+  );
+
+  if (!hasMigrations) {
+    return MigrationAction.Migrate;
+  }
+
+  const refreshedSource = await Index.fetchInfo(callCluster, alias);
+
+  if (!refreshedSource.aliases[alias]) {
+    return MigrationAction.Migrate;
+  }
+
+  return diffMapping(refreshedSource.mappings, dest.mappings);
+}
+
+/**
+ * Applies the latest mappings to the index.
+ */
+async function patchSourceMappings(context: Context): Promise<MigrationResult> {
+  const { callCluster, log, source, dest } = context;
+
+  log.info(`Patching ${source.indexName} mappings`);
+
+  await Index.putMappings(callCluster, source.indexName, dest.mappings);
+
+  return { status: 'patched' };
+}
+
+/**
+ * Performs an index migration if the source index exists, otherwise
+ * this simply creates the dest index with the proper mappings.
+ */
+async function migrateIndex(context: Context): Promise<MigrationResult> {
+  const startTime = Date.now();
+  const { callCluster, alias, source, dest, log } = context;
+
+  log.info(`Creating index ${dest.indexName}.`);
+
+  await Index.createIndex(callCluster, dest.indexName, dest.mappings);
+
+  await migrateSourceToDest(context);
+
+  log.info(`Pointing alias ${alias} to ${dest.indexName}.`);
+
+  await Index.claimAlias(callCluster, dest.indexName, alias);
+
+  const result: MigrationResult = {
+    status: 'migrated',
```

Comment:
I don't have a strong opinion about this, but why use special string values for "status" rather than an enum like you did for MigrationAction?

Reply:
Mainly because I have integration tests in plain old JavaScript, so it was easier to make this a string.